### PR TITLE
feat: Support nested locale keys.

### PR DESCRIPTION
**Pull request by huacnlee**
Example:

```js
// config/locales/en.js
module.exports = {
  model: {
    user: {
      name: 'Real Name',
      createdAt: 'Joined At'
    }
  }
};
```

```js
ctx.__('model.user.name');
ctx.__('model.user.createdAt');
```

The before:

```js
// config/locales/en.js
module.exports = {
  'model.user.name': 'Real Name',
  'model.user.createdAt', 'Joined At'
};
```

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -103,7 +103,17 @@ module.exports = function (app, options) {
 
     var locale = this.__getLocale();
     var resource = resources[locale] || {};
-    var text = resource[key] || key;
+
+    // Try to match nested keys, example: model.user.fields.title
+    var keys = key.split('.');
+    var nestedText = resource;
+    for (var i = 0; i < keys.length; i++) {
+      if (nestedText === undefined) {
```
Comment (hotoo) [inline index.js]:
```diff
+ nestedText = null;
```

**Comment by hotoo** [comment]
Add benchmark test.

File: index.js [hunk 2]
```diff
@@ -103,7 +103,17 @@ module.exports = function (app, options) {
...
+        break;
+      }
+      nestedText = nestedText[keys[i]];
+    }
+    var text = resource[key] || nestedText || key;
```
Comment (hotoo) [inline index.js]:
```js
function getNestedKey(root, keys) { ... }

var text = resource[key] || getNestedKey(resource, key) || key;
```

**Comment by fengmk2** [comment]
image

将你本地的 node_modules 删除了，就能重新了。 @popomore 感觉是你新提交的那个 muk 出 bug 了。

File: index.js [hunk 3]
```diff
@@ -103,7 +103,17 @@ module.exports = function (app, options) {
...
+
+    // Try to match nested keys, example: model.user.fields.title
+    var keys = key.split('.');
+    var nestedText = resource;
+    for (var i = 0; i < keys.length; i++) {
+      if (nestedText === undefined) {
+        break;
+      }
+      nestedText = nestedText[keys[i]];
+    }
```
Comment (hotoo) [inline index.js]:
```js
if (!isString(nestedText) {
  nestedText = null;
}
```

File: test/locales/zh_TW.json [hunk 4]
```diff
@@ -1,4 +1,12 @@
 {
   "Email": "郵箱",
-  "Hello %s, how are you today?": "%s，今天過得如何？"
+  "Hello %s, how are you today?": "%s，今天過得如何？",
+  "model": {
+    "user": {
+      "fields": {
+        "name": "姓名",
+        "gender": "性別"
```
Comment (hotoo) [inline test/locales/zh_TW.json]: Not good enough, what about one more thing:

```json
{
  "model.user.fields": {
    "name": "姓名",
    "gender": "性别"
  }
}
```

File: index.js [hunk 5]
```diff
@@ -103,7 +103,17 @@ module.exports = function (app, options) {
 
     var locale = this.__getLocale();
     var resource = resources[locale] || {};
-    var text = resource[key] || key;
+
+    // Try to match nested keys, example: model.user.fields.title
```
Comment (hotoo) [inline index.js]: Do this on load localization data maybe better.

app.context[functionName] need not to change.

File: test/locales/zh_TW.json (hunk 4 above)
Comment (huacnlee) [inline test/locales/zh_TW.json]: More complex example in real world:

```js
{
  model: {
    hint: {
      user: {
        name: '要求 2 - 10 个字符',
        age: '必须是整数'
      },
      post: {
        title: '标题长度要求在 5 - 100 字之间'
      }
    },
    placeholder: {
      user: {
        email: 'your@email.com',
        homepage: 'http://youdomain.com'
      }
    },
    user: {
      name: '姓名',
      age: '年龄'
    },
    post: {
      title: '标题',
      body: '正文'
    }
  }
}
```

Comment (hotoo) [inline test/locales/zh_TW.json]: Support both of them.

Comment (hotoo) [inline test/locales/zh_TW.json]: AND,,,

```
model.hint.user.name
model.hint.user.age
model.hint.post.title
model.placeholder.user.email
model.placeholder.user.homepage
model.user.name
model.user.name
model.post.title
model.post.body
```

is not bad.

**Comment by huacnlee** [comment]
Too slow, I will try the another way.

```
  3 tests completed.

  direct read a key x 88,459,126 ops/sec ±0.91% (98 runs sampled)
  by nested         x  5,632,453 ops/sec ±3.74% (87 runs sampled)
```

**Comment by hotoo** [comment]
Do this on load localization data maybe better.

app.context[functionName] not need to change.

**Comment by hotoo** [comment]
TOO SLOW.

direct read a key is fast 15 times than by nested.

**Comment by huacnlee** [comment]
~= 5,632 ops/ms.

That mean it just spend 1ms even have 5000+ locale call in a page.

**Comment by huacnlee** [comment]
But I will try that faster way, I just say that not slow.

**Comment by popomore** [comment]
Relate https://github.com/fent/node-muk/pull/5

**Comment by popomore** [comment]
I had unpublished muk@0.3.3

**Comment by fengmk2** [comment]
> ~= 5,632 ops/ms.

**ms** .... is ms!!!

+1 fast enough!

**Comment by fengmk2** [comment]
1.4.0

**Comment by hotoo** [comment]
最重要的一个建议没采纳，居然 :u5408: 了。

**Pull request merged**